RealmSwift/Query.swift

```diff
@@ -0,0 +1,1075 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2021 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+import Foundation
+import Realm
+
+/// Enum representing an option for `String` queries.
+public enum StringOptions {
+    /// A case-insensitive search.
+    case caseInsensitive
+    /// Search ignores diacritic marks.
+    case diacriticInsensitive
+}
+
+private enum QueryExpression {
```

Comment:
This representation of an expression makes all of the code working with it very complicated. Building a tree rather than a flat array with some complicated non-end insertions to mimic a tree structure would be a lot less error-prone. It's also not obvious to me that we need our own data type for this at all. Is there anything which would be difficult to do if the builder functions simply assembled a NSPredicate directly?

Reply:
There are a few cases which make assembling an NSPredicate directly tricky, the two main ones are subqueries and some Map expressions. I'm happy to take a look and see if I can achieve anything with the tree approach and if that leads to better NSPredicate generation.

Reply:
I'm doing a bit of refactoring replacing `QueryExpression` with NSExpression and NSCompoundPredicate etc.